runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java

```diff
@@ -58,6 +86,107 @@ public SamzaRunner(SamzaPipelineOptions options) {
+            }
+
+            if (SamzaRunnerOverrideConfigs.isPortableMode(options)) {
+              if (jobBundleFactory == null) {
+                try {
+                  final long waitTimeoutMs =
+                      SamzaRunnerOverrideConfigs.getControlClientWaitTimeoutMs(options);
+                  final InstructionRequestHandler instructionHandler =
+                      controlClientPool
+                          .getSource()
+                          .take(SAMZA_WORKER_ID, Duration.ofMillis(waitTimeoutMs));
+                  final EnvironmentFactory environmentFactory =
+                      environment -> RemoteEnvironment.forHandler(environment, instructionHandler);
+                  // TODO: use JobBundleFactoryBase.WrappedSdkHarnessClient.wrapping
+                  jobBundleFactory =
+                      SingleEnvironmentInstanceJobBundleFactory.create(
+                          environmentFactory, fnDataServer, fnStateServer);
+                } catch (Exception e) {
+                  throw new RuntimeException(
+                      "Running samza in Beam portable mode but failed to create job bundle factory",
+                      e);
+                }
+                executionContext.setJobBundleFactory(jobBundleFactory);
+              }
+            }
+
+            context.setUserContext(executionContext);
+          }
+
+          @Override
+          public void close() {
+            closeFnServer(fnControlServer);
+            fnControlServer = null;
+            closeFnServer(fnDataServer);
+            fnDataServer = null;
+            closeFnServer(fnStateServer);
+            fnStateServer = null;
+          }
+        });
+  }
+
+  private void setUpFnApiServer() {
+    controlClientPool = MapControlClientPool.create();
+    ExecutorService dataExecutor = Executors.newCachedThreadPool();
+    try {
+      fnControlServer =
+          GrpcFnServer.allocatePortAndCreateFor(
+              FnApiControlClientPoolService.offeringClientsToPool(
+                  controlClientPool.getSink(), () -> SAMZA_WORKER_ID),
+              ServerFactory.createWithPortSupplier(
+                  () -> SamzaRunnerOverrideConfigs.getFnControlPort(options)));
+
+      fnDataServer =
+          GrpcFnServer.allocatePortAndCreateFor(
+              GrpcDataService.create(dataExecutor, OutboundObserverFactory.serverDirect()),
+              ServerFactory.createDefault());
+
+      fnStateServer =
+          GrpcFnServer.allocatePortAndCreateFor(
+              GrpcStateService.create(), ServerFactory.createDefault());
+    } catch (Exception e) {
+      LOG.error("Failed to set up fn api servers", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  SamzaPipelineResult runPortablePipeline(RunnerApi.Pipeline pipeline) {
+    final SamzaExecutionContext executionContext = new SamzaExecutionContext();
+    setUpFnApiServer();
```

Comment:
setupFnApiServer() needs to be folded inside setupContextManager() since it needs to be set up only when the stream graph is created during runtime. Currently it was done during submission time.

Reply:
As discussed, put a TODO here as it's already done in 1.0 migrate